x-pack/filebeat/input/streaming/websocket.go

```diff
@@ -82,48 +89,151 @@ func (s *websocketStream) FollowStream(ctx context.Context) error {
 	}
 
 	// websocket client
-	headers := formHeader(s.cfg)
-	c, resp, err := websocket.DefaultDialer.DialContext(ctx, url, headers)
+	c, resp, err := connectWebSocket(ctx, s.cfg, url, s.log)
+	handleConnectionResponse(resp, s.log)
+	if err != nil {
+		s.metrics.errorsTotal.Inc()
+		s.log.Errorw("failed to establish websocket connection", "error", err)
+		return err
+	}
+	// ensures this is the last connection closed when the function returns
+	defer c.Close()
+
+	for {
+		select {
+		case <-ctx.Done():
+			s.log.Debugw("context cancelled, closing websocket connection")
+			c.Close()
+			return ctx.Err()
+		default:
+			_, message, err := c.ReadMessage()
+			if err != nil {
+				s.metrics.errorsTotal.Inc()
+				if isRetryableError(err) {
+					s.log.Debugw("websocket connection encountered an error, attempting to reconnect...", "error", err)
+					// close the old connection and reconnect
+					c.Close()
+					// since c is already a pointer, we can reassign it to the new connection and the defer will still handle it
+					c, resp, err = connectWebSocket(ctx, s.cfg, url, s.log)
+					handleConnectionResponse(resp, s.log)
+					if err != nil {
+						s.metrics.errorsTotal.Inc()
+						s.log.Errorw("failed to reconnect websocket connection", "error", err)
+						return err
+					}
```

Comment:
This new websocket connection also needs to be closed. The defer statement will hold a reference to the function itself, so replacing `c` won't affect the deferred call.
```suggestion
					c, resp, err = connectWebSocket(ctx, s.cfg, url, s.log)
					handleConnectionResponse(resp, s.log)
					if err != nil {
						s.metrics.errorsTotal.Inc()
						s.log.Errorw("failed to reconnect websocket connection", "error", err)
						return err
					}
					defer c.Close()
```

Here is an example from the Go Playground I used to test it: https://go.dev/play/p/nBvZRoBYTBc

Reply:
Then maybe we can just wrap the defer in a separate function call and defer the wrapped function. Changing c in this scenario will ideally reflect in the deferred statement. I'm hesitant to defer from within a loop as it's not considered a good design pattern when it comes to go.